### ISSUE TITLE
some torrents are corrupted after editing or passing thru retrackers plugin 

### ISSUE THREAD

**Issue by beelze** [Novik/ruTorrent#2246]
At first, I've noticed some torrents «can't be added» – they are appeared for a short time but quickly disappeared after. After some investigation I've realized that was caused by retrackers plugin (even if announcers not been touched at all – I've changed a code to check this). Almost the same happening with edit – torrent disappearing after pressing ok button. Looking into a code (I'm not a php coder, though) I've realized that torrents is loaded from file, processed and sent back after removing original torrent:
```
                        $torrent = new Torrent( $fname );  
....
                       $eReq = new rXMLRPCRequest( new rXMLRPCCommand("d.erase", $hash ) );                             
                       if($eReq->success())                                                                             
                       {                                                                                                
                           $label = rawurldecode($req->val[3]);                                                         
                           rTorrent::sendTorrent($torrent, $isStart, false, $req->val[4], $label, false, false, false,  
                                                 array(getCmd("d.set_custom3")."=1") );                                 
```
My guess is something in rTorrent class is incorrectly processing metadata (only for some edge cases). As I've said already, I'm not a php coder so I'm not able to fix it. I've attached such an «edge case» .torrent here
[a2.torrent.zip](https://github.com/Novik/ruTorrent/files/7798954/a2.torrent.zip)
.

**Comment by beelze** [comment]
UPD: Forgot to mention – issue is relevant to files count in metadata (about ~10K in example .torrent). I've tried to recreate metadata from downloaded files and problem was disappeared after files count has been lowered to ~9K. (it's not an exact numbers, just for illustration)

**Comment by stickz** [comment]
Yes, I believe this is a problem with rTorrent. I had a similar issue when trying to "remove torrents and delete data". The files in the directory would not actually delete in some edge cases. 

I looked on the rTorrent handbook for more information about `d.erase` since this seems to be related to your code snippet.
https://rtorrent-docs.readthedocs.io/en/latest/cmd-ref.html#term-d-erase

I would recommend trying to running `chmod -R 777` on your` .session` directory for rTorrent and going from there.

**Comment by beelze** [comment]
It is not permission-related issue. As I said before, I was able to create 2 almost identical .torrents, from same source. The first one can't be loaded, the second one loaded without errors. The only difference is files count (I've removed one small directory, ~70 small files)

**Comment by beelze** [comment]
I believe I got close to a problem (please do not blame me if I'm wrong anywhere – I'm not a PHP guy at all):
1. retrackers plugin creates `Torrent` object from file in .session by hash
2. alters its `announce_list` (in memory!)
3. deletes «old» torrent and sends altered object: `rTorrent::sendTorrent($torrent, ... `

please note the following code snippet (php/rtorrent.php, rTorrent::sendTorrent()):
```
const RTORRENT_PACKET_LIMIT = 1572864;     
...
            $raw_value = base64_encode($torrent->__toString());                   
            $filename = is_object($fname) ? $torrent->getFileName() : $fname;     
            if((strlen($raw_value)<self::RTORRENT_PACKET_LIMIT) || is_null($filename) || !isLocalMode())  
            {                                                                                             
                $cmd = new rXMLRPCCommand( $isStart ? 'load_raw_start' : 'load_raw' );                    
                $cmd->addParameter($raw_value,"base64");                                                  
                if(!is_null($filename) && !$saveTorrent)                                                  
                    @unlink($filename);                                                                   
            }                                                                                             
            else                                                                                          
            {                                                                                             
                $cmd = new rXMLRPCCommand( $isStart ? 'load_start' : 'load' );                            
                $cmd->addParameter($filename);                                                            
            }                                                                                             
```
`$filename` is always not null (because object was created from file) and `isLocalMode()` is true (in my case). Thus, if bencoded data size is greater `RTORRENT_PACKET_LIMIT`, it will `load`|`load_start` .torrent from file, **ignoring in-memory changes**. In other hand, is size less than `RTORRENT_PACKET_LIMIT`, it will `load_raw_start`|`load_raw` bencoded data, which is proper behaviour.

I guess there is no bug in retrackers code – it simply assumes that `sendTorrent()` will send changed data. In contrary, `sendTorrent()` behaviour is unpredictable because it can ignore all the changes for big .torrents. 

I fixed this issue in by saving .torrent before sending (rutrackers/update.php):
```
$torrent->save($fname);
rTorrent::sendTorrent($torrent, ...
```
but it's a obviously dirty hack. I believe that `sendTorrent()` should use something like dirty flag and save bencoded data to temporary file is needed before issuing `load*` to rtorrent. Sorry but I have no PHP skills and therefore unable to create proper patch.

**Comment by stickz** [comment]
Could you try a modification for me, since you're able to re-simulate this problem? 

Add this function to the Torrent class in `php/Torrent.php` file.
```
	public function setFileName($filename = null)
	{
		$this->filename = is_null( $filename ) ? $this->info['name'] . '.torrent' : $filename;
	}
```

Then call it before sending the torrent in `plugins/retrackers/update.php`.
```
$torrent->setFileName($fname);
rTorrent::sendTorrent($torrent, ...
```

I don't want to write the torrent file to the disk before sending it out. The way it streams the file creates unnecessary overhead.

**Comment by beelze** [comment]
Please correct me if I'm wrong, but as far as I can see it will change nothing in my case because it won't set local `$filename` inside `sendTorrent()` to null (and, therefore, `sendTorrent()` will call `load*` FILE if size > `RTORRENT_PACKET_LIMIT` as before)

If we want to **not** stream the file, we should remove the size check. But I believe if we ignore blindly this condition, we will ruin something.
 
In other words, if we want to keep `load*` functionality in `sendTorrent()`, we should save file first if needed (but where?)

**Comment by beelze** [comment]
the real question is: **why** and **when** we should use `load*` instead `load_raw*`? What's the limit for sending metadata in-band, using `load_raw*`?

**Comment by stickz** [comment]
Can you please just give this a try? It will take maybe 15 minutes. The torrent class might be messing around with the filename.

**Comment by Novik** [comment]
@stickz 

> public function setFileName($filename = null)

This will not work.
We need a real .torrent file on the disk for usage of load_start/load.

> the real question is: why and when we should use load* instead load_raw*

load_raw_start/load_raw can't be used when
1) we have rtorrent on other host or
2) we have too big torrent file. 
In this case we should save torrent to the temporary file first.

**Comment by beelze** [comment]
> We should save torrent to the temporary file first.

I believe this should be implemented inside `sendTorrent()`

**Comment by Novik** [comment]
@beelze 

I made corresponding changes to the master branch, check result, please.

**Comment by beelze** [comment]
@Novik 
Rougly backported to 3.10. As for now, the problem seems solved.
